onnxruntime/core/framework/sparse_tensor.cc

```diff
@@ -3,31 +3,441 @@
+
+const void* SparseTensor::IndicesStart(int64_t values_bytes) const {
+  if (p_data_ != nullptr) {
+    return reinterpret_cast<const uint8_t*>(p_data_) + Roundup(values_bytes);
+  }
+  return nullptr;
+}
+
+void* SparseTensor::IndicesStart(int64_t values_bytes) {
+  if (p_data_ != nullptr) {
+    return reinterpret_cast<uint8_t*>(p_data_) + Roundup(values_bytes);
+  }
+  return nullptr;
+}
+
+int64_t SparseTensor::RequiredAllocationSize() const noexcept {
+  if (p_data_ != nullptr) {
+    // Can be zero for zero matrices
+    assert(buffer_size_ >= 0);
+    return buffer_size_;
+  } else {
+    auto data_size = values_.SizeInBytes();
+    int64_t index_size = 0;
+    for (const auto& t : format_data_) {
+      index_size += t.SizeInBytes();
+    }
+    return CalculateRequiredBufferSize(data_size, index_size);
+  }
+}
+
+SparseTensor::SparseTensor(MLDataType elt_type, const TensorShape& dense_shape,
+                           const TensorShape& values_shape, void* values_data,
+                           const OrtMemoryInfo& location)
+    : SparseTensor() {
+  dense_shape_ = dense_shape;
+  ml_data_type_ = elt_type->AsPrimitiveDataType();
+  location_ = location;
+  values_ = Tensor(elt_type, values_shape, values_data, location_);
+}
+
+SparseTensor::SparseTensor(MLDataType elt_type, const TensorShape& dense_shape,
                            std::shared_ptr<IAllocator> allocator)
-    : values_(elt_type, TensorShape({static_cast<int64_t>(nnz)}), allocator),
-      indices_(DataTypeImpl::GetType<int64_t>(),
-               TensorShape({static_cast<int64_t>(nnz), static_cast<int64_t>(shape.NumDimensions())}),
-               allocator),
-      shape_(shape) {}
+    : SparseTensor() {
+  dense_shape_ = dense_shape;
+  ml_data_type_ = elt_type->AsPrimitiveDataType();
+  allocator_ = std::move(allocator);
+  location_ = allocator_->Info();
+}
+
+SparseTensor::SparseTensor() noexcept
+    : format_(SparseFormat::kUndefined),
+      dense_shape_(),
+      ml_data_type_(nullptr),
+      allocator_(),
+      location_(),
+      p_data_(nullptr),
+      buffer_size_(0),
+      values_(),
+      format_data_() {
+}
+
+SparseTensor::SparseTensor(SparseTensor&& o) noexcept : SparseTensor() {
+  *this = std::move(o);
+}
+
+SparseTensor& SparseTensor::operator=(SparseTensor&& o) noexcept {
+  ReleaseBuffer();
+  format_ = o.format_;
+  dense_shape_ = std::move(o.dense_shape_);
+  ml_data_type_ = o.ml_data_type_;
+  allocator_ = std::move(o.allocator_);
+  location_ = std::move(o.location_);
+  std::swap(p_data_, o.p_data_);
+  std::swap(buffer_size_, o.buffer_size_);
+  values_ = std::move(o.values_);
+  format_data_ = std::move(o.format_data_);
+  return *this;
+}
+
+SparseTensor::~SparseTensor() {
+  ReleaseBuffer();
+}
+
+Status SparseTensor::AllocateBuffer(int64_t buffer_size, size_t num_values) {
+  if (buffer_size > 0) {
+    SafeInt<size_t> b_size_t(buffer_size);
+    ORT_ENFORCE(b_size_t.Ref() > (num_values * ml_data_type_->Size()),
```

Comment:
>Ref [](http://example.com/codeflow?start=25&length=3)

was looking up Ref() and noticed a comment said that static_cast is better.
[https://github.com/dcleblanc/SafeInt/blob/1619e36802d928e6d9af0ab38bcc1269ff77f1ad/SafeInt.hpp#L5992](https://github.com/dcleblanc/SafeInt/blob/1619e36802d928e6d9af0ab38bcc1269ff77f1ad/SafeInt.hpp#L5992)
but here, are we circumventing the safety check when doing the comparison?